examples/drivers/windows/ottmp/hdlc.cpp

```diff
@@ -0,0 +1,343 @@
+}
+
+ULONG
+HdlcComputeEncodedLength(
+    _In_reads_bytes_(DecodedBufferLength) 
+         PUCHAR         DecodedBuffer,
+    _In_ ULONG          DecodedBufferLength
+    )
+{
+    USHORT Fcs = kInitFcs;
+    ULONG Length = 2; // Start and end sequence
+    
+    for (ULONG i = 0; i < DecodedBufferLength; i++)
+    {
+        Length += HdlcComputeEncodedByteLength(DecodedBuffer[i]);
+        Fcs = UpdateFcs(Fcs, DecodedBuffer[i]);
+    }
+    
+    Fcs ^= 0xffff;
+    
+    Length += HdlcComputeEncodedByteLength(Fcs & 0xff);
+    Length += HdlcComputeEncodedByteLength(Fcs >> 8);
+
+    return Length;
+}
+
+bool
+HdlcEncodeByte(
+    _In_    UCHAR       aByte,
+    _Inout_ ULONG&      Offset,
+    _In_    ULONG       EncodedBufferLength,
+    _Out_writes_bytes_(EncodedBufferLength) 
+            PUCHAR      EncodedBuffer
+    )
+{
+    if (hdlc_byte_needs_escape(aByte))
+    {
+        if (Offset + 2 > EncodedBufferLength) return false;
+
+        EncodedBuffer[Offset++] = HdlcEscapeSequence;
+        EncodedBuffer[Offset++] = (aByte ^ 0x20);
+    }
+    else
+    {
+        if (Offset + 1 > EncodedBufferLength) return false;
+
+        EncodedBuffer[Offset++] = aByte;
+    }
+
+    return true;
+}
+
+_Success_(return == true)
+bool
+HdlcEncodeBuffer(
+    _In_reads_bytes_(DecodedBufferLength) 
+         PUCHAR         DecodedBuffer,
+    _In_ ULONG          DecodedBufferLength,
+    _Out_writes_bytes_(EncodedBufferLength) 
+         PUCHAR         EncodedBuffer,
+    _In_ ULONG          EncodedBufferLength
+    )
+{
+    USHORT Fcs = kInitFcs;
+    ULONG Offset = 0;
+    bool Complete = false;
+
+    if (Offset + 1 > EncodedBufferLength) goto error;
+    EncodedBuffer[Offset++] = HdlcFlagSequence;
+
+    for (ULONG i = 0; i < DecodedBufferLength; i++)
+    {
+        UCHAR aByte = DecodedBuffer[i];
+
+        if (!HdlcEncodeByte(aByte, Offset, EncodedBufferLength, EncodedBuffer)) {
+            goto error;
+        }
+
+        Fcs = UpdateFcs(Fcs, aByte);
+    }
+    
+    Fcs ^= 0xffff;
+    
+    if (!HdlcEncodeByte(Fcs & 0xff, Offset, EncodedBufferLength, EncodedBuffer) ||
+        !HdlcEncodeByte(Fcs >> 8,   Offset, EncodedBufferLength, EncodedBuffer)) {
+        goto error;
+    }
+    
+    if (Offset + 1 > EncodedBufferLength) goto error;
+    EncodedBuffer[Offset++] = HdlcFlagSequence;
+
+    NT_ASSERT(Offset == EncodedBufferLength);
+    Complete = true;
+
+error:
+
+    return Complete;
+}
+
+enum HdlcDecodeState
+{
+    kStateNoSync = 0,
+    kStateSync,
+    kStateEscaped,
+};
+
+_Success_(return == true)
+bool
+HdlcDecodeBuffer(
+    _In_reads_bytes_(*EncodedBufferLength) 
+            PUCHAR      EncodedBuffer,
+    _Inout_ PULONG      EncodedBufferLength,
+    _Inout_ PULONG      DecodedBufferLength,
+    _Out_writes_bytes_opt_(*DecodedBufferLength) 
+            PUCHAR      DecodedBuffer,
+    _Out_   bool*       IsGood
+    )
+{
+    UCHAR byte;
+    HdlcDecodeState state = kStateNoSync;
```

Comment:
It wasn't immediately clear to me how this function works if the state isn't preserved across calls. Does it just not update `EncodedBufferLength` if it hasn't hit the frame marker?